File: web/src/utils/dates.ts
/**
 * Format a Date as YYYY-MM-DD using local timezone.
 * This avoids the UTC offset issue where .toISOString() can return
 * the wrong date for users in negative UTC offsets late in the evening.
 */
function formatLocalDate(d: Date): string {
  const year = d.getFullYear();
  const month = String(d.getMonth() + 1).padStart(2, "0");
  const day = String(d.getDate()).padStart(2, "0");
  return `${year}-${month}-${day}`;
}

/**
 * Format a string like "YYYY-MM-DD" to a short display format e.g. "Jun 1".
 * Month abbreviations can be localized by passing the translated month list
 * (e.g. from the i18n dictionary); it defaults to English.
 */
function formatShortDisplay(dateStr: string, months: string[] = [
  "Jan", "Feb", "Mar", "Apr", "May", "Jun",
  "Jul", "Aug", "Sep", "Oct", "Nov", "Dec",
]): string {
  const d = new Date(dateStr + "T00:00:00");
  return `${months[d.getMonth()]} ${d.getDate()}`;
}

/**
 * Get ISO date strings (YYYY-MM-DD) for a given calendar week
 * (Monday to Sunday), using local timezone. Used by the weekly Planner grid,
 * which is anchored to fixed Monday–Sunday columns.
 *
 * @param weekOffset - Offset relative to the current week:
 *   0 (default) = current week,
 *  -1 = previous week,
 *  +1 = next week, etc.
 */
export function getWeekDates(weekOffset = 0): string[] {
  const dates: string[] = [];
  const now = new Date();
  const dayOfWeek = now.getDay();
  const monday = new Date(now);
  monday.setDate(now.getDate() - ((dayOfWeek + 6) % 7) + weekOffset * 7);
  for (let i = 0; i < 7; i++) {
    const d = new Date(monday);
    d.setDate(monday.getDate() + i);
    dates.push(formatLocalDate(d));
  }
  return dates;
}

/**
 * Get ISO date strings (YYYY-MM-DD) for a rolling 7-day window.
 *
 * Unlike getWeekDates (anchored to Monday), this window is anchored to today:
 * offset 0 returns the last 7 days ending today, with today as the final
 * column. This keeps habit streaks flowing continuously between calendar
 * weeks — on Monday you still see the previous week's completions instead of
 * a hard reset to a blank Monday–Sunday grid.
 *
 * @param offset - Offset relative to the current window:
 *   0 (default) = last 7 days ending today,
 *  -1 = the 7 days before that,
 *  +1 = the 7 days after that (future — usually not shown).
 */
export function getRollingWindowDates(offset = 0): string[] {
  const dates: string[] = [];
  const today = new Date();
  const end = new Date(today);
  end.setDate(today.getDate() + offset * 7);
  const start = new Date(end);
  start.setDate(end.getDate() - 6);
  for (let i = 0; i < 7; i++) {
    const d = new Date(start);
    d.setDate(start.getDate() + i);
    dates.push(formatLocalDate(d));
  }
  return dates;
}

/**
 * Get today's date as an ISO string (YYYY-MM-DD) using local timezone.
 */
export function getToday(): string {
  return formatLocalDate(new Date());
}

/**
 * Format a week's date range into a compact display string.
 * Same-month: "1–7 Jun"
 * Cross-month: "29 Jun – 5 Jul"
 *
 * @param monthNames Optional localized month abbreviations (length 12). Falls
 *   back to English when omitted.
 */
export function formatWeekRange(weekDates: string[], monthNames?: string[]): string {
  const start = formatShortDisplay(weekDates[0], monthNames);
  const end = formatShortDisplay(weekDates[6], monthNames);
  // Check if both dates are in the same month
  const startMonth = start.split(" ")[0];
  const endMonth = end.split(" ")[0];
  if (startMonth === endMonth) {
    return `${start.split(" ")[1]}–${end.split(" ")[1]} ${startMonth}`;
  }
  return `${start} – ${end}`;
}

/**
 * Normalize a time string from the backend (e.g. "09:00:00") to the
 * HH:MM format expected by HTML <input type="time"> elements.
 * If the input is invalid or empty, returns an empty string.
 */
export function sanitizeTime(t: string | null | undefined): string {
  if (!t) return "";
  const parts = t.split(":");
  if (parts.length < 2) return "";
  return `${parts[0]}:${parts[1]}`;
}

File: mobile/app/src/main/java/com/pudimproductivity/utils/Dates.kt
package com.pudimproductivity.utils

import java.time.DayOfWeek
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import java.time.temporal.TemporalAdjusters

/**
 * Returns 7 date strings (YYYY-MM-DD) for the week starting Monday,
 * offset by [weekOffset] weeks (0 = current week, -1 = last week, etc.).
 */
fun getWeekDates(weekOffset: Int = 0): List<String> {
    val today = LocalDate.now()
    val monday = today.with(TemporalAdjusters.previousOrSame(DayOfWeek.MONDAY))
        .plusWeeks(weekOffset.toLong())
    return (0..6).map { monday.plusDays(it.toLong()).format(DateTimeFormatter.ISO_LOCAL_DATE) }
}

/**
 * Returns today's date as YYYY-MM-DD.
 */
fun getToday(): String {
    return LocalDate.now().format(DateTimeFormatter.ISO_LOCAL_DATE)
}

/**
 * Returns 7 ISO date strings (YYYY-MM-DD) for a rolling 7-day window anchored
 * to today (mirrors the web's `getRollingWindowDates`).
 *
 * Unlike [getWeekDates] (fixed Monday–Sunday), offset 0 returns the last 7 days
 * ending today, with today as the final column. This keeps habit streaks flowing
 * continuously between calendar weeks — on Monday you still see the previous
 * week's completions instead of a hard reset to a blank Monday–Sunday grid.
 *
 * @param offset Offset relative to the current window:
 *   0 (default) = last 7 days ending today,
 *  -1 = the 7 days before that,
 *  +1 = the 7 days after that (future — usually not shown).
 */
fun getRollingWindowDates(offset: Int = 0): List<String> {
    val end = LocalDate.now().plusWeeks(offset.toLong())
    val start = end.minusDays(6)
    return (0..6).map { start.plusDays(it.toLong()).format(DateTimeFormatter.ISO_LOCAL_DATE) }
}

/**
 * Returns the date `dayOffset` days from today as YYYY-MM-DD (used by the
 * single-day Planner, where 0 = today, -1 = yesterday, +1 = tomorrow).
 */
fun getDate(dayOffset: Int = 0): String {
    return LocalDate.now().plusDays(dayOffset.toLong()).format(DateTimeFormatter.ISO_LOCAL_DATE)
}

/**
 * Given a list of 7 ISO date strings (Mon–Sun), returns a human-friendly
 * range like "Mar 15–21". Month abbreviations can be localized via [monthNames]
 * (defaults to English).
 */
fun formatWeekRange(
    dates: List<String>,
    monthNames: List<String> = listOf("Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec")
): String {
    if (dates.size < 2) return ""
    val start = formatShortDate(dates.first(), monthNames)
    val end = formatShortDate(dates.last(), monthNames)
    // Same month → "Mar 15–21", different months → "Mar 28–Apr 3"
    val startMonth = dates.first().substring(5, 7)
    val endMonth = dates.last().substring(5, 7)
    return if (startMonth == endMonth) {
        "${monthNames[startMonth.toInt() - 1]} ${start.substringAfter(" ")}–${end.substringAfter(" ")}"
    } else {
        "$start – $end"
    }
}

private fun formatShortDate(isoDate: String, monthNames: List<String>): String {
    val date = LocalDate.parse(isoDate)
    return "${monthNames[date.monthValue - 1]} ${date.dayOfMonth}"
}